### PR TITLE
[inductor] write cpp code with encoding utf-8

### PR DESCRIPTION
**Pull request by xuhancn**
Windows is different to Linux, each Windows version with different language pack have different code page. 
Inductor on Windows will write the genarated cpp code with its code page, and it should occured un-decode character failed.

For this situlation, Microsoft suggest to use Unicode to instead of a specific code page. Ref: https://learn.microsoft.com/en-us/windows/win32/intl/code-page-identifiers

Changes:
1. Use `utf-8` as encoder for cpp code.
2. It only change encode for cpp code, but not for binary type. binary type is for AoT binary context.

It works on https://github.com/pytorch/pytorch/issues/122094#issuecomment-2299592942. 

cc @peterjc123 @mszhanyi @skyline75489 @nbcsm @iremyux @Blackhex @jgong5 @mingfeima @XiaobingSuper @sanchitintel @ashokei @jingxu10 @voznesenskym @penguinwu @EikanWang @Guobing-Chen @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @ipiszy @yf225 @chenyang78 @kadeng @muchulee8 @ColinPeppler @amjames @desertfire @chauhang

### CODE REVIEW

File: torch/_inductor/codecache.py [hunk 1]
```diff
@@ -454,8 +458,12 @@ def write_atomic(
         path.parent.mkdir(parents=True, exist_ok=True)
     tmp_path = path.parent / f".{os.getpid()}.{threading.get_ident()}.tmp"
     write_mode = "w" if isinstance(content, str) else "wb"
-    with tmp_path.open(write_mode) as f:
-        f.write(content)
+    if encode_utf_8:
+        with tmp_path.open(write_mode, encoding="utf-8") as f:
```
Comment (desertfire) [inline torch/_inductor/codecache.py:462]: Cleaner to use a single context but move the condition to `encoding`.

**Review by desertfire — APPROVED** [review]

File: torch/_inductor/codecache.py (hunk 1 above)
Comment (xuhancn) [reply torch/_inductor/codecache.py]: @desertfire Good suggestion, done.

**Review by jgong5 — APPROVED** [review]

**Review by jansel — APPROVED** [review]

**Pull request closed without merging**